Linux ip.Addr and ip.Link plugins

gcmoreira opened the pull request:
This PR adds the `linux.ip.Addr` and `linux.ip.Link` plugins. 

### linux.ip.Addr:
 - It displays information about all network interfaces including network namespace, interface index, interface name, mac address, promiscuous status, IPv4/IPv6 address, prefix, scope type and state.

Example output:
```shell
$ python3 ./vol.py -r pretty \
  -f ram-6.2.0-26 \
  linux.ip.Addr
Volatility 3 Framework 2.5.2
Formatting...0.00               Stacking attempts finished                 
  |      NetNS | Index | Interface |               MAC | Promiscuous |                       IP | Prefix | Scope Type |   State
* | 4026531840 |     1 |        lo | 00:00:00:00:00:00 |       False |                127.0.0.1 |      8 |       host | UNKNOWN
* | 4026531840 |     1 |        lo | 00:00:00:00:00:00 |       False |                      ::1 |    128 |       host | UNKNOWN
* | 4026531840 |     2 |     ens32 | 00:0c:29:7f:8b:ab |       False |           172.16.141.130 |     24 |     global |      UP
* | 4026531840 |     2 |     ens32 | 00:0c:29:7f:8b:ab |       False | fe80::20c:29ff:fe7f:8bab |     64 |       link |      UP
* | 4026531840 |     3 |     ens33 | 00:0c:29:7f:8b:b5 |       False |          192.168.249.129 |     24 |     global |      UP
* | 4026531840 |     3 |     ens33 | 00:0c:29:7f:8b:b5 |       False | fe80::20c:29ff:fe7f:8bb5 |     64 |       link |      UP
``` 

### linux.ip.Link (by @eve-mem )
 - It displays information about all network devices configuration including network namespace,  interface name, mac address, state, MTU, Qdisc, Qlen and flags.
```shell
$ python3 ./vol.py -r pretty \
  -f ram-6.2.0-26 \
  linux.ip.Link
Volatility 3 Framework 2.5.2
Formatting...0.00               Stacking attempts finished                 
  |         NS | Interface |               MAC |   State |   MTU |    Qdisc | Qlen |                           Flags
* | 4026531840 |        lo | 00:00:00:00:00:00 | UNKNOWN | 65536 |  noqueue | 1000 |            LOOPBACK,LOWER_UP,UP
* | 4026531840 |     ens32 | 00:0c:29:7f:8b:ab |      UP |  1500 | fq_codel | 1000 | BROADCAST,LOWER_UP,MULTICAST,UP
* | 4026531840 |     ens33 | 00:0c:29:7f:8b:b5 |      UP |  1500 | fq_codel | 1000 | BROADCAST,LOWER_UP,MULTICAST,UP
```

Both plugins were tested with the following linux kernel versions:
- 2.6.32-74.142
- 3.10.0-862_el7 
- 3.11.0-26
- 4.4.0-210
- 4.13.0-46
- 4.15.0-45
- 4.18.0-10
- 5.3.0-76
- 5.8.0-53
- 5.19.0-50
- 6.2.0-26

See the full test case suite output:
[vol3_linux_ip_addr_output.txt](https://github.com/volatilityfoundation/volatility3/files/14074292/vol3_linux_ifconfig_output.txt)

[vol3_linux_ip_link_output.txt](https://github.com/volatilityfoundation/volatility3/files/14138737/vol3_linux_ip_link.txt)

gcmoreira committed:
Linux - Added linux.ifconfig.Ifconfig plugin

gcmoreira committed:
Fix exception. Although it will be auto-instantiated it's better to explicitily use the exception instance

gcmoreira commented:
Hey @ikelos I'm aware of this #1029 excellent contribution from @eve. We've already discussed this, and we agreed to collaborate, combining our efforts into one.
I will borrow some part of his code, for instance, to display the interface state and use the address conversion from the vol3 helpers instead from the python socket module.
Regarding the plugin name, I like the @eve-mem idea of named it as the new linux commands `ip address`. That also allows to continue developing other plugins like 'ip link' (@eve-mem already have done that) and other future commands. 
My only concern is that probably for users coming from vol2 they would like to see this as `ifconfig`. Any preference?

eve-mem commented:
@gcmoreira - looks really cool, I need to look over it properly. I certainly like pulling out the prefix into it's own column compared with https://github.com/volatilityfoundation/volatility3/pull/1029 - that makes it easier to work with programmatically later. 

+1 vote to renaming the plugins to linux.ip.addr etc to match the new linux commands. There are already a bunch of commands that were effectively renamed between vol2 and vol3 already.

gcmoreira committed:
Plugin renamed. Added net iface status and other improvements from @eve #1029

* IP address conversion via renderers.coversion.*
* Use MAC address internal size instead of hardcoded.
* Read NET_DEVICE_FLAGS from enumeration

gcmoreira committed:
Minor fixes

gcmoreira committed:
Fix: Explicit returns mixed with implicit (fall through) returns

gcmoreira committed:
Merge branch 'volatilityfoundation:develop' into linux_ifconfig_plugin

ikelos commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -1262,16 +1264,41 @@ def get_mac_address(self):
+
+    def _get_net_device_flag_value(self, name):
+        """Return the net_deivce flag value based on the flag name"""
+        return self._get_flag_choices()[name]
```
I'd be happier if this were a get with a default value?  Having said that, it's such a short function, perhaps it's not worth having at all (given it's private so can't be used by anyone else?)

gcmoreira replied:
Default value and error managed. 
I prefer to keep it as a function, even though it's short and private. This way, it serves the purpose of documenting what it does and increases the code's readability. I would let the language to implement the bytecode inline if needed.
Anyway, if you still think it should be inline, let me know, and I will change it.

ikelos commented on volatility3/framework/constants/linux/__init__.py:
```diff
@@ -291,5 +291,37 @@
+}
+
+# RFC 2863 operational status. Kernels >= 2.6.17. See IF_OPER_* in include/uapi/linux/if.h
+IF_OPER_STATES = (
```
Should these be `Enum`s rather than just a list of strings?  If may potentially lead to typos?

gcmoreira replied:
Done. This is used by value, so I don't think at the moment being an Enum offers any usability advantage. Although I reckon it may look better to Pythonic eyes ;)

ikelos commented on volatility3/framework/constants/linux/__init__.py:
```diff
@@ -291,5 +291,37 @@
-IFF_PROMISC = 0x100
+# Only for kernels < 3.15 when the net_device_flags enum didn't exist
+# ref include/uapi/linux/if.h
+NET_DEVICE_FLAGS = {
```
This also looks like it could be a python Enum?

gcmoreira replied:
hm here changing this to an Enum makes the code a bit complicated... have a look at how it's used in `_get_flag_choices()` and `_get_net_device_flag_value()` to support both kernels <3.15 and >=3.15.

ikelos reviewed (commented):
Generally looks good, just a couple of minor points to check please.  5:)

gcmoreira commented:
Thanks @ikelos. I will have a look at this soon. I also plan to include the ip.link plugin from @eve-mem in this PR, that's why I marked it as a draft.

In the meantime, could you have a look at the black formatter checks? I'm confused, as far as I understand it's failing on files which are not from this PR

eve-mem commented:
@gcmoreira - it looks nice to me. I'll run it against my collection of samples when I can. Can I help adding in `ip.Link` at all?

eve-mem commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -1224,6 +1236,221 @@ def get_inode(self):
+        """
+        if self.operstate >= len(IF_OPER_STATES):
+            vollog.warning(f"Invalid net_device operational state '{self.operstate}'")
+            return "INVALID"
```
Rather than "INVALID" as a string would an `UnparsableValue` from `volatility3/volatility3/framework/renderers` be better here?

ikelos commented:
It is, or was, github updated to black 24.1.0, so I went over the while
codebase, but it doesn't seem to get that the merge *target* updated, so
returning the task still results in the same output.  5:S. Basically if you
add a new commit, it should do a proper refresh and either pass or show you
how the new black failed it.  Hard to tell if there's still something that
needs doing, but just a trivial commit will help and I figured since it was
in draft it would get added to... 5:)

On Mon, 29 Jan 2024, 00:55 gcmoreira, ***@***.***> wrote:

> Thanks @ikelos <https://github.com/ikelos>. I will have a look at this
> soon. I also plan to include the ip.link plugin from @eve-mem
> <https://github.com/eve-mem> in this PR, that's why I marked it as a
> draft.
>
> In the meantime, could you have a look at the black formatter checks? I'm
> confused, as far as I understand it's failing on files which are not from
> this PR
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/volatilityfoundation/volatility3/pull/1079#issuecomment-1913792237>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALIZVIJNUJJARGODCOUILTYQ3XOXAVCNFSM6AAAAABBTT6DKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMJTG44TEMRTG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

gcmoreira committed:
Fix docstring typos

gcmoreira committed:
Convert IF_OPER_STATES to enum

gcmoreira committed:
Manage net_device flag default value & error

gcmoreira commented:
> It is, or was, github updated to black 24.1.0, so I went over the while codebase, but it doesn't seem to get that the merge *target* updated, so returning the task still results in the same output. 5:S. Basically if you add a new commit, it should do a proper refresh and either pass or show you how the new black failed it.

Yep, that did the trick, thanks

gcmoreira commented:
> @gcmoreira - it looks nice to me. I'll run it against my collection of samples when I can. Can I help adding in `ip.Link` at all?

@eve-mem sure, go ahead with that ;)

gcmoreira committed:
Add test for linux.ip.Addr and linux-sample-1.bin

gcmoreira committed:
Python os.path module precisely does that by checking the current platform and, based on it, utilizes either the posixpath or ntpath modules

gcmoreira committed:
Linux: Add the linux.ip.Link plugin by @eve-mem

- On top of the @eve-mem, I've added the queue length field to mimic the ip link command.
- Furthermore, I've included some functions to export the network device flags exactly as they are presented to userland

gcmoreira committed:
Add linux.ip.Link test using linux-sample-1.bin image

gcmoreira committed:
Remove lowercase matching from linux.ip.Addr test

gcmoreira committed:
Fix issue with net namespace id and improve code

gcmoreira commented:
@ikelos This is now ready for review. Added testcases for both plugins

gcmoreira commented:
Hi @ikelos, this is still awaiting a review. Maybe @atcuno  can have a look? :pray:

eve-mem commented:
Looks like there are some merge conflicts with this now.

ikelos commented:
@gcmoreira could you or @atcuno resolve the merge conflicts and @atcuno could you please get this reviewed, given it's marked for the parity release?

ikelos commented:
Also, since this wants to make a lot of changes to the LinuxUtilities class, could we try splitting it out into a utilities class, similar to the tracing stuff (#1564), so that it can be individually versioned...

ikelos committed:
Merge branch 'develop' into linux_ifconfig_plugin

ikelos commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -1832,7 +2173,8 @@ def get_family(self):
 
     def get_type(self):
-        return linux_constants.SOCK_TYPES.get(self.sk_type, "")
+        return linux_constants.
```
Urgh, looks like I broke this, sorry!  5:S

ikelos replied:
```suggestion
        return linux_constants.SOCK_TYPES.get(self.sk_type, "")
```

ikelos commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -1832,7 +2173,8 @@ def get_family(self):
     def get_type(self):
-        return linux_constants.SOCK_TYPES.get(self.sk_type, "")
+        return linux_constants.
+      .get(self.sk_type, "")
```
```suggestion

```

ikelos committed:
Apply suggestions from code review

Hopefully I can fix up my breakage...

ikelos commented on volatility3/framework/constants/linux/__init__.py:
```diff
@@ -282,6 +282,51 @@
+    DORMANT = 5
+    UP = 6
 
 class ELF_IDENT(IntEnum):
```
```suggestion
class ELF_IDENT(enum.IntEnum):
```

ikelos commented on volatility3/framework/constants/linux/__init__.py:
```diff
@@ -282,6 +282,51 @@
+
+
+# Kernels >= 2.6.17. See IF_OPER_* in include/uapi/linux/if.h
+class IF_OPER_STATES(Enum):
```
```suggestion
class IF_OPER_STATES(enum.Enum):
```

ikelos commented on volatility3/framework/constants/linux/__init__.py:
```diff
@@ -415,4 +459,4 @@ class TaintFlag:
     - taint_flag kernel struct
     - taint_flags kernel constant
-"""
+"""
```
```suggestion
"""

```

ikelos commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -1793,6 +1803,337 @@ def get_inode(self):
+            return "host"
+        elif (self.scope & linux_constants.IFA_LINK) != 0:
+            return "link"
+        elif (self.scope & linuc_constants.IFA_SITE) != 0:
```
```suggestion
        elif (self.scope & linux_constants.IFA_SITE) != 0:
```

ikelos commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -1793,6 +1803,337 @@ def get_inode(self):
+            return
+
+        # 'if_list' member was added to 'inet6_ifaddr' type in kernels 3.0
+        for inet6_ifaddr in self.addr_list.to_list(inet6_ifaddr_symname, "if_list"):
```
```suggestion
        yield from self.addr_list.to_list(inet6_ifaddr_symname, "if_list"):
```

ikelos commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -1793,6 +1803,337 @@ def get_inode(self):
+
+        # 'if_list' member was added to 'inet6_ifaddr' type in kernels 3.0
+        for inet6_ifaddr in self.addr_list.to_list(inet6_ifaddr_symname, "if_list"):
+            yield inet6_ifaddr
```
```suggestion
```

ikelos committed:
Apply suggestions from code review

Try to patch up ruff failures through suggestions.

ikelos commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -1793,6 +1803,336 @@ def get_inode(self):
+            return
+
+        # 'if_list' member was added to 'inet6_ifaddr' type in kernels 3.0
+        yield from self.addr_list.to_list(inet6_ifaddr_symname, "if_list"):
```
```suggestion
        yield from self.addr_list.to_list(inet6_ifaddr_symname, "if_list")
```

ikelos committed:
Update volatility3/framework/symbols/linux/extensions/__init__.py

My bad, little typo when trying to fix it last time

ikelos commented:
Ok, this one I can't fix with a suggestion, it's too far away in the code, just needs `import enum` (or reverting my module name changes, and `from enum import Enum, IntEnum, Flags`).  Sorry @gcmoreira, I tried!

ikelos commented on volatility3/framework/symbols/linux/extensions/__init__.py:
```diff
@@ -1793,6 +1803,336 @@
+            return
+
+        # 'if_list' member was added to 'inet6_ifaddr' type in kernels 3.0
+        yield from self.addr_list.to_list(inet6_ifaddr_symname, "if_list"):
```
Fixed now

ikelos committed:
Merge branch 'feature/linux_ifconfig_plugin' into linux_ifconfig_plugin
